enshrine-esdt-safe/src/from_sovereign/transfer_tokens.rs

```diff
@@ -36,13 +39,99 @@ pub trait TransferTokensModule:
             sc_panic!("Operation is not registered");
         }
 
-        let minted_operation_tokens = self.mint_tokens(&operation.tokens);
-        let operation_tuple = OperationTuple {
-            op_hash: operation_hash,
-            operation,
-        };
+        let (wegld_amount, remaining_tokens) = self.verify_operation_tokens_for_issue_fee(
+            &operation.data.op_sender,
+            operation.tokens.clone(),
+        );
+
+        if remaining_tokens.len() > 0 {
+            let minted_operation_tokens = self.mint_tokens(&remaining_tokens);
+            let operation_tuple = OperationTuple {
+                op_hash: operation_hash.clone(),
+                operation: operation.clone(),
+            };
+
+            self.distribute_payments(
+                hash_of_hashes.clone(),
+                operation_tuple,
+                minted_operation_tokens,
+            );
+        }
+
+        self.execute_bridge_operation_event(hash_of_hashes, operation_hash);
+    }
+
+    #[endpoint(registerTokens)]
+    #[payable("*")]
+    fn register_tokens(&self, tokens: MultiValueEncoded<TokenIdentifier>) {
+        let call_payment = self.call_value().single_esdt().clone();
+        let wegld_identifier = self.wegld_identifier().get();
+
+        require!(
+            call_payment.token_identifier == wegld_identifier,
+            "WEGLD is the only token accepted as register fee"
+        );
+
+        require!(
+            call_payment.amount == DEFAULT_ISSUE_COST * tokens.len() as u64,
+            "WEGLD fee amount is not met"
+        );
+
+        for token_id in tokens {
+            self.register_token(token_id);
+        }
+    }
+
+    fn verify_operation_tokens_for_issue_fee(
+        &self,
+        sender: &ManagedAddress<Self::Api>,
+        tokens: ManagedVec<OperationEsdtPayment<Self::Api>>,
+    ) -> ManagedVec<OperationEsdtPayment<Self::Api>> {
+        let first_payment = tokens.get(0);
+        let is_first_payment_wegld = self.is_wegld(&first_payment.token_identifier);
+
+        if !is_first_payment_wegld {
+            for token in tokens.iter() {
+                if !self.was_token_registered(&token.token_identifier)
```

Comment:
The condition should be `if token has prefix && is not registered then return ERROR` and at line 101 `return tokens`